README.md
<img src="https://raw.githubusercontent.com/esnet/react-timeseries-charts/master/src/website/packages/charts/logo.png" alt="logo" width="160px"/>

# React Timeseries Charts

[![Build Status](https://travis-ci.org/esnet/react-timeseries-charts.svg)](https://travis-ci.org/esnet/react-timeseries-charts) [![npm version](https://badge.fury.io/js/react-timeseries-charts.svg)](https://badge.fury.io/js/react-timeseries-charts)

![Zooming](https://github.com/esnet/react-timeseries-charts/blob/master/zooming.gif)

This library contains a set of modular charting components used for building flexible interactive charts. It was built for React from the ground up, specifically to visualize timeseries data and network traffic data in particular. Low level elements are constructed using d3, while higher level composability is provided by React. Charts can be stacked as rows, overlaid on top of each other, or any combination, all in a highly declarative manner.

The library is used throughout the public facing [ESnet Portal](http://my.es.net).

Current features of the library include:

 * Declarative layout of charts using JSX
 * Interactivity, including pan and zoom
 * Add new chart types or overlays
 * Multiple axis, multiple rows
 * Line, area, scatter, bar and boxplot charts
 * Brushing
 * Legends
 * Overlay markers

Please browse the [examples](http://software.es.net/react-timeseries-charts) for a feel for the library.

Getting started
---------------

This charts library is intended to be installed with [npm](https://www.npmjs.com/) and the built into your project with a tool like [Webpack](https://webpack.github.io/). It expects React to be present, as well as our TimeSeries abstraction library, [pond.js](http://software.es.net/pond). More on this below. To install:

    npm install react-timeseries-charts pondjs --save

Once installed, you can import the necessary components from the library:

    import { Charts, ChartContainer, ChartRow, YAxis, LineChart } from "react-timeseries-charts";

Then we construct our chart in the `render()` function of our component. For a simple example we create a chart with two line charts on it, specified in JSX:

    <ChartContainer timeRange={series1.timerange()} width={800}>
        <ChartRow height="200">
            <YAxis id="axis1" label="AUD" min={0.5} max={1.5} width="60" type="linear" format="$,.2f"/>
            <Charts>
                <LineChart axis="axis1" series={series1}/>
                <LineChart axis="axis2" series={series2}/>
            </Charts>
            <YAxis id="axis2" label="Euro" min={0.5} max={1.5} width="80" type="linear" format="$,.2f"/>
        </ChartRow>
    </ChartContainer>

At the outer most layer, we add a `<ChartContainer>` which contains our time range for the x-axis. All charts within a ChartContainer share the same x-axis. In this case we get the TimeRange from the TimeSeries itself, but you could specify one yourself. You also need to provide a width for the chart, or wrap the chart in a `<Resizable>` component and that will inject a width for you.

For the next layer of the layout we make a `<ChartRow>`. We can have multiple charts stacked on top of each other by using more than one row. In this case we just have one row. Each row has a specific height in the layout, so we specify that as 200px high here.

Next up we want to put something in our row. Rows contain two parts:
1. a central flexible sized area in which charts can be added and  
2. axes on either the left or right of the central area.

This central area is surrounded in the JSX by the `<Charts>` tag. Each chart in this area is composited on top of each other. In this case we are adding two `<LineChart>`s, one for each of our timeseries. As a result they will be drawn on top of each other. (Note that as of v0.9, it is also possible to draw multiple channels of a TimeSeries as multiple line charts). For scaling each chart will reference an axis that will define the scale as well as display that scale visually as the y-axis.

Finally, we specify the axes that the charts reference. These go either before or after the `<Charts>` group, depending on if you want the axis before (to the left) or after the charts (to the right). You can specify any number of axes on either side. For each `<YAxis>` you specify the `id` so that a chart can reference it, the `label` you want displayed alongside the axis and, importantly, the scale using `min` and `max`. You can also specify the `type` of the axis (`linear`, `log`, etc), a non-default `width` and the `format`.

For more details, please see the API docs.

Data format
-----------

This charting library is built on the ESnet timeseries library called [pond.js](http://software.es.net/pond). Pond.js was created to provide a common representation for things like time series, time ranges, events and other structures as well as to implement common operations on those structures.

For the purpose of using the charting library, you need to create a `TimeSeries` object as your time series, which will be rendered by the chart code. Constructing a TimeSeries is pretty simple. You just need to pass it an object that contains, at a minimum, the column names as an array, and a list of points.

For example:

    import { TimeSeries, TimeRange } from "pondjs";

    const data = {
        name: "traffic",
        columns: ["time", "in", "out"],
        points: [
            [1400425947000, 52, 41],
            [1400425948000, 18, 45],
            [1400425949000, 26, 49],
            [1400425950000, 93, 81],
            ...
        ]
    };

    const timeseries = new TimeSeries(data);


Generally `columns` must have 'time' as the first column (see the Pond.js docs for other options), and `points` is of the format:

    [[time, value], [time, value], ...]

Note that there are several ways to construct a TimeSeries. The above is what we call our wire format.

To specify a ChartContainer you also need to provide the time range to plot on the x-axis. This is done with a pond.js `TimeRange`. You could either get the TimeRange from the TimeSeries directly:

    var timerange = timeseries.timerange()

Or construct a new one:

    var timerange = new TimeRange([begin, end]);

where `begin` and `end` are times (Javascript Dates, usually).

How you manage this is application specific. We often have pages where the current time range being shown is kept in the `this.state` of the component and is updated during pan and zoom, both via interaction with the charts or via other controls on the page.

Developing
----------

The repo contains the examples website. This is very helpful in developing new functionality. Within a cloned repo, you first need to run:

    yarn install

This will install the development dependencies into your node_modules directory.

You can then start up the test server:

    npm run start-website

Then, point your browser to:

[http://localhost:3000/](http://localhost:3000/)

Before submitting a Pull Request, run:
    
    npm run build
    
In case you want to demonstrate an issue with the library via an example, you can use [CodeSandBox](https://codesandbox.io/s/).

A simple BarChart example has been built here - [https://codesandbox.io/s/3j0540mo5](https://codesandbox.io/s/3j0540mo5)

Supported Browsers
-------

The most recent release of Firefox or Chrome is recommended.

License
-------

This code is distributed under a BSD style license, see the LICENSE file for complete information.

Copyright
---------

ESnet's React Timeseries Charts, Copyright (c) 2015 - present, The Regents of the University of California, through Lawrence Berkeley National Laboratory (subject to receipt of any required approvals from the U.S. Dept. of Energy). All rights reserved.

If you have questions about your rights to use or distribute this software, please contact Berkeley Lab's Technology Transfer Department at TTD@lbl.gov.

NOTICE. This software is owned by the U.S. Department of Energy. As such, the U.S. Government has been granted for itself and others acting on its behalf a paid-up, nonexclusive, irrevocable, worldwide license in the Software to reproduce, prepare derivative works, and perform publicly and display publicly. Beginning five (5) years after the date permission to assert copyright is obtained from the U.S. Department of Energy, and subject to any subsequent five (5) year renewals, the U.S. Government is granted for itself and others acting on its behalf a paid-up, nonexclusive, irrevocable, worldwide license in the Software to reproduce, prepare derivative works, distribute copies to the public, perform publicly and display publicly, and to permit others to do so.


CHANGES.md
# Changelog

## v0.16.1

> May 17, 2019

-   Add ability to specify the title of ChartRow [PR #359](https://github.com/esnet/react-timeseries-charts/pull/359)
-   Add customized text styling in ValueList.js [#355](https://github.com/esnet/react-timeseries-charts/issues/355)
-   Fix YAxis component label not updating after initial render [PR #351](https://github.com/esnet/react-timeseries-charts/pull/351)
-   Fix ChartRow so as to not call setState after being unmounted [#246](https://github.com/esnet/react-timeseries-charts/issues/246)
-   Fix issue where the context menu triggers drag as well [#307](https://github.com/esnet/react-timeseries-charts/issues/307)
-   Fix for Chrome 73 default passive event listeners / Fix scrolling to zoom in a ChartContainer that also scrolls the entire page [#375](https://github.com/esnet/react-timeseries-charts/issues/375)
-   Fix for intermittently click event not working on EventChart [PR #369](https://github.com/esnet/react-timeseries-charts/pull/369)
-   Fix YAxis rerender on tickCount/min/max change [#371](https://github.com/esnet/react-timeseries-charts/issues/371)
-   Add options modify the vertical axis margin and vertical EventMarker infobox offset [PR #377](https://github.com/esnet/react-timeseries-charts/pull/377)
-   Fixes the documentation page on Safari [PR #387](https://github.com/esnet/react-timeseries-charts/issues/387)
-   Create a template for new issues

## v0.15.5

> Nov 1, 2018

Bug Fixes:

-   Update to create-react-app V2.0
-   Add note to documentation that columns can't have periods
-   Export BandChart component
-   Add an example to style BandChart component
-   Add a check for undefined value in the ScatterChart component

## v0.15.4

> Oct 26, 2018

-   Removes flexbox-react dependency [#310](https://github.com/esnet/react-timeseries-charts/issues/310)
-   Fixes Resizable component binding to `this` [#329](https://github.com/esnet/react-timeseries-charts/issues/329)

## v0.15.3

> Aug 15, 2018

-   Add feature to add style prop to `LabelAxis` component
-   Refactor `AreaChart` code

Bug fixes:

-   Corrects the wrong order for testing style type (Fixes [#277](https://github.com/esnet/react-timeseries-charts/issues/277))
-   Correctly stacking in area charts when some points are bad (Fixes [#284](https://github.com/esnet/react-timeseries-charts/issues/284))
-   Adds prop to hide TimeAxis (Fixes [#295](https://github.com/esnet/react-timeseries-charts/issues/295))
-   Adds function argument while styling barcharts (Fixes [#291](https://github.com/esnet/react-timeseries-charts/issues/291))
-   Adds a prop `marginBottom` to the `Legend` component (Fixes [#243](https://github.com/esnet/react-timeseries-charts/issues/243))
-   Doesn't draw a `BarChart` when the height of the rectangle is 0 (Fixes [#294](https://github.com/esnet/react-timeseries-charts/issues/294))
-   Removes warnings on climate example (Fixes [#257](https://github.com/esnet/react-timeseries-charts/issues/257))
-   Exposes x, y and time value to onBackgroundClick method in `ChartContainer` (Fixes [#296](https://github.com/esnet/react-timeseries-charts/issues/296))
-   TimeMarker can be custom styled (Fixes [#304](https://github.com/esnet/react-timeseries-charts/issues/304))
-   Ability to style legend (Fixes [#112](https://github.com/esnet/react-timeseries-charts/issues/112))

## v0.15.0

> May 22, 2018

-   Adds `YAxis` and `TimeAxis` styling of all elements. This is a breaking change, but now allows much more flexible styling. See the the DDoS example. In [#267](https://github.com/esnet/react-timeseries-charts/pull/267).
-   `YAxis` now supports horizontal grid lines. This only applies to axes which are adjacent to the chart (Fixes #244). In [#267](https://github.com/esnet/react-timeseries-charts/pull/267).
-   `TimeAxis` now supports angled labels [#229](https://github.com/esnet/react-timeseries-charts/pull/229)
-   You can now add a title to the chart by passing one into the `ChartContainer` (Fixes #9)
-   `AreaChart`s now have an option to break at data gaps (i.e. ignore `nulls` and `NaNs`) in [#247](https://github.com/esnet/react-timeseries-charts/pull/247)
-   `ChartContainer` now has a margin prop that will surround the chart with empty space. This is useful to avoid clipped `TimeAxis` labels [#229](https://github.com/esnet/react-timeseries-charts/pull/229)
-   Add drag to zoom feature by [@iabw](https://github.com/iabw) in [#252](https://github.com/esnet/react-timeseries-charts/pull/252)
-   Current `TimeAxis` scale function exposed as second argument to `onTrackerChanged()` by [@ethanroday](https://github.com/ethanroday) in [#238](https://github.com/esnet/react-timeseries-charts/pull/238)
-   Allow `YAxis` yscale by [@ethanroday](https://github.com/ethanroday) in [#251](https://github.com/esnet/react-timeseries-charts/pull/251)

Bug fixes:

-   `TimeAxis` and `YAxis` now updates to additional prop changes (Fixes #259, Fixes #255 )
-   Ease required props on `TimeRangeMarker` (Fixes #263)
-   `ChartRow` is better protected against null children [#253](https://github.com/esnet/react-timeseries-charts/pull/253)
-   Fix issues with using webpack and hot loader in [#248](https://github.com/esnet/react-timeseries-charts/pull/248)

## v0.14.0

> March 2, 2018

-   Add MultiBrush component and update traffic example to implement the same [@barbaragomes](https://github.com/barbaragomes) in [#224](https://github.com/esnet/react-timeseries-charts/pull/224)
-   Add a crosshairs custom chart to the currency demo
-   Ability to ignore null points in ScatterChart [#226]
-   Exposes hover x,y events in ChartContainer handleMouseMove [#199]
-   Added a control for TimeAxis tick count [#220]
-   Control minimum BarChart height [#218]
-   Fixed LineChart example text [#215]
-   Baseline has better control over label vertical positioning [#216]
-   Legend horizontal layout now wraps automatically [#202]
-   Legend can now be laid out in a column (see weather example)

lib/yaxis.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _react = require("react");

var _react2 = _interopRequireDefault(_react);

var _reactDom = require("react-dom");

var _reactDom2 = _interopRequireDefault(_reactDom);

var _d3Axis = require("d3-axis");

var _d3Format = require("d3-format");

var _d3Selection = require("d3-selection");

var _d3Transition = require("d3-transition");

var _d3Ease = require("d3-ease");

require("./yaxis.css");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 *  Copyright (c) 2015, The Regents of the University of California,
 *  through Lawrence Berkeley National Laboratory (subject to receipt
 *  of any required approvals from the U.S. Dept. of Energy).
 *  All rights reserved.
 *
 *  This source code is licensed under the BSD-style license found in the
 *  LICENSE file in the root directory of this source tree.
 */

var MARGIN = 0;

function scaleAsString(scale) {
    if (scale.domain && scale.range) {
        return scale.domain().toString() + "-" + scale.range().toString();
    } else {
        return "scale-" + Math.random;
    }
}

/**
 * The YAxis widget displays a vertical axis to the left or right
 * of the charts. A YAxis always appears within a `ChartRow`, from
 * which it gets its height and positioning. You can have more than
 * one axis per row.
 *
 * ![YAxis](https://raw.githubusercontent.com/esnet/react-timeseries-charts/master/docs/yaxis.png "YAxis")
 *
 * Here's a simple YAxis example:
 *
 * ```js
 * <YAxis id="price-axis" label="Price (USD)" min={0} max={100} width="60" type="linear" format="$,.2f"/>
 * ```
 *
 * Visually you can control the axis `label`, its size via the `width`
 * prop, its `format`, and `type` of scale (linear).
 *
 * Each axis also defines a scale through a `min` and `max` prop. Charts
 * may then refer to the axis by by citing the axis `id` in their `axis`
 * prop. Those charts will then use the axis scale for their y-scale.
 *
 * Here is an example of two line charts that each have their own axis:
 *
 * ```js
 * <ChartContainer timeRange={audSeries.timerange()}>
 *     <ChartRow height="200">
 *         <YAxis id="aud" label="AUD" min={0.5} max={1.5} width="60" type="linear" format="$,.2f"/>
 *         <Charts>
 *             <LineChart axis="aud" series={audSeries} style={audStyle}/>
 *             <LineChart axis="euro" series={euroSeries} style={euroStyle}/>
 *         </Charts>
 *         <YAxis id="euro" label="Euro" min={0.5} max={1.5} width="80" type="linear" format="$,.2f"/>
 *     </ChartRow>
 * </ChartContainer>
 * ```
 *
 *  Note that there are two `<YAxis>` components defined here, one before
 *  the `<Charts>` block and one after. This defines that the first axis will
 *  appear to the left of the charts and the second will appear after the charts.
 *  Each of the line charts uses its `axis` prop to identify the axis ("aud" or "euro")
 *  it will use for its vertical scale.
 */
exports.default = _react2.default.createClass({

    displayName: "YAxis",

    getDefaultProps: function getDefaultProps() {
        return {
            id: "yaxis", // id referred to by the chart
            align: "left", // left or right of the chart
            min: 0, // range
            max: 1,
            type: "linear", // linear, log, or power
            absolute: false, // Display scale always positive
            format: ".2s", // Format string for d3.format
            labelOffset: 0, // Offset the label position
            transition: 100, // Axis transition time
            width: 80,
            style: {
                labelColor: "#8B7E7E", // Default label color
                labelWeight: 100,
                labelSize: 12
            }
        };
    },


    propTypes: {

        /**
         * A name for the axis which can be used by a chart to reference the axis.
         */
        id: _react2.default.PropTypes.string.isRequired,

        /**
         * The label to be displayed alongside the axis.
         */
        label: _react2.default.PropTypes.string,

        /**
         * The scale type: linear, power, or log.
         */
        type: _react2.default.PropTypes.oneOf(["linear", "power", "log"]),

        /**
         * Minium value, which combined with "max", define the scale of the axis.
         */
        min: _react2.default.PropTypes.number.isRequired,

        /**
         * Maxium value, which combined with "min,"" define the scale of the axis.
         */
        max: _react2.default.PropTypes.number.isRequired,

        /**
         * The transition time for moving from one scale to another
         */
        transition: _react2.default.PropTypes.number,

        /**
         * The width of the axis
         */
        width: _react2.default.PropTypes.oneOfType([_react2.default.PropTypes.string, _react2.default.PropTypes.number]),

        /**
         * d3.format for the axis labels. e.g. `format="$,.2f"`
         */
        format: _react2.default.PropTypes.string,

        /**
         * If the chart should be rendered to with the axis on the left or right.
         * If you are using the axis in a ChartRow, you do not need to provide this.
         */
        align: _react2.default.PropTypes.string

    },

    renderAxis: function renderAxis(align, scale, width, absolute, fmt) {
        var yformat = (0, _d3Format.format)(fmt);
        var axisGenerator = void 0;
        var axis = align === "left" ? _d3Axis.axisLeft : _d3Axis.axisRight;
        if (this.props.type === "linear" || this.props.type === "power") {
            if (this.props.height <= 200) {
                axisGenerator = axis(scale).ticks(5).tickFormat(function (d) {
                    if (absolute) {
                        return yformat(Math.abs(d));
                    } else {
                        return yformat(d);
                    }
                });
            } else {
                axisGenerator = axis(scale).tickFormat(function (d) {
                    if (absolute) {
                        return yformat(Math.abs(d));
                    } else {
                        return yformat(d);
                    }
                });
            }
        } else if (this.props.type === "log") {
            axisGenerator = axis().scale(scale).ticks(10, ".2s");
        }

        // Remove the old axis from under this DOM node
        (0, _d3Selection.select)(_reactDom2.default.findDOMNode(this)).selectAll("*").remove();

        // Add the new axis
        var x = align === "left" ? width - MARGIN : 0;
        var labelOffset = align === "left" ? this.props.labelOffset - 50 : 40 + this.props.labelOffset;
        var classed = this.props.classed ? this.props.classed : "";
        var axisClass = "yaxis " + classed;
        var axisLabelClass = "yaxis-label " + classed;

        this.axis = (0, _d3Selection.select)(_reactDom2.default.findDOMNode(this)).append("g").attr("transform", "translate(" + x + ",0)").attr("class", axisClass).call(axisGenerator).append("text").text(this.props.label).attr("transform", "rotate(-90)").attr("class", axisLabelClass).attr("y", labelOffset).attr("dy", ".71em").attr("text-anchor", "end").style("fill", this.props.style.labelColor).style("font-family", this.props.style.labelFont || "\"Goudy Bookletter 1911\", sans-serif\"").style("font-weight", this.props.style.labelWeight || 100).style("font-size", this.props.style.labelSize ? this.props.style.width + "px" : "12px");
    },
    updateAxis: function updateAxis(align, scale, width, absolute, type, fmt) {
        var yformat = (0, _d3Format.format)(fmt);
        var axis = align === "left" ? _d3Axis.axisLeft : _d3Axis.axisRight;
        var axisGenerator = void 0;
        if (type === "linear" || type === "power") {
            if (this.props.height <= 200) {
                axisGenerator = axis(scale).ticks(5).tickFormat(function (d) {
                    if (absolute) {
                        return yformat(Math.abs(d));
                    } else {
                        return yformat(d);
                    }
                });
            } else {
                axisGenerator = axis(scale).tickFormat(function (d) {
                    if (absolute) {
                        return yformat(Math.abs(d));
                    } else {
                        return yformat(d);
                    }
                });
            }
        } else if (type === "log") {
            axisGenerator = axis(scale).ticks(10, ".2s");
        }

        var t = (0, _d3Transition.transition)().duration(750).ease(_d3Ease.easeSinOut);

        (0, _d3Selection.select)(_reactDom2.default.findDOMNode(this)).select(".yaxis").transition(t).call(axisGenerator);
    },
    componentDidMount: function componentDidMount() {
        this.renderAxis(this.props.align, this.props.scale, +this.props.width, this.props.absolute, this.props.format);
    },
    componentWillReceiveProps: function componentWillReceiveProps(nextProps) {
        var scale = nextProps.scale;
        var align = nextProps.align;
        var width = nextProps.width;
        var absolute = nextProps.absolute;
        var format = nextProps.format;
        var type = nextProps.type;

        if (scaleAsString(this.props.scale) !== scaleAsString(scale) || this.props.type !== nextProps.type) {
            this.updateAxis(align, scale, width, absolute, type, format);
        }
    },
    shouldComponentUpdate: function shouldComponentUpdate() {
        return false;
    },
    render: function render() {
        return _react2.default.createElement("g", null);
    }
});